test/integration/util/util.go

```diff
@@ -123,6 +123,10 @@ func StartFakePVController(ctx context.Context, clientSet clientset.Interface, i
 	pvInformer := informerFactory.Core().V1().PersistentVolumes()
 
 	syncPV := func(obj *v1.PersistentVolume) {
+		if ctx.Err() != nil {
+			// We can stop...
```

Comment:
I'd prefer to see there is some err or warning here to say that context is cancelled, or some error is occurred, unless this is already print to somewhere else.

Reply:
"Context cancelled" is normal and not an unexpected error. Therefore it doesn't need to be logged.

Reply:
Instead, we should probably be canceling the event handlers when the context is cancelled.

Reply:
> how do we know what's happened?

It doesn't matter whether the context got cancelled or timed out. In both cases these functions should return. At best it's a debug message at high verbosity, not an error.

> Instead, we should probably be canceling the event handlers when the context is cancelled.

If we don't cancel the `ctx` that the fake controller is using, then there's no guarantee that the code that it is running terminates in a timely manner.

Reply:
Stopping the informer factory when the context is canceled is [already done](https://github.com/kubernetes/kubernetes/blob/9e0569f2ed3934060fabe51be4e15232bbea3877/test/integration/util/util.go#L100).

But even with that it is possible that a sync is running or gets started when the context is canceled, which leads to useless "error while getting" error logs during shutdown.

Looking at this again, I realized that my if check wasn't in quite the right place: it must come after an API call failed, otherwise there's a race. Fixed.